subsys/bluetooth/audio/has.c

```diff
@@ -0,0 +1,1330 @@
+		if (!atomic_test_bit(client->flags, CLIENT_FLAG_ATT_MTU_VALID)) {
+			return BT_GATT_ERR(BT_ATT_ERR_INSUFFICIENT_RESOURCES);
+		}
+
+		ret = handle_write_preset_name(client, &buf);
+		break;
+#endif /* CONFIG_BT_HAS_PRESET_NAME_DYNAMIC */
+
+	case BT_HAS_OP_SET_ACTIVE_PRESET:
+		if (!atomic_test_bit(client->flags, CLIENT_FLAG_CP_IND_ENABLED)) {
+			return BT_GATT_ERR(BT_ATT_ERR_CCC_IMPROPER_CONF);
+		}
+
+		ret = handle_set_active_preset(client, &buf, false);
+		break;
+
+	case BT_HAS_OP_SET_NEXT_PRESET:
+		ret = handle_set_next_preset(client, &buf, false);
+		break;
+
+	case BT_HAS_OP_SET_PREV_PRESET:
+		ret = handle_set_prev_preset(client, &buf, false);
+		break;
+
+#if defined(CONFIG_BT_HAS_PRESET_SYNC_SUPPORT)
+	case BT_HAS_OP_SET_ACTIVE_PRESET_SYNC:
+		if (!atomic_test_bit(client->flags, CLIENT_FLAG_CP_IND_ENABLED)) {
+			return BT_GATT_ERR(BT_ATT_ERR_CCC_IMPROPER_CONF);
+		}
+
+		ret = handle_set_active_preset(client, &buf, true);
+		break;
+
+	case BT_HAS_OP_SET_NEXT_PRESET_SYNC:
+		ret = handle_set_next_preset(client, &buf, true);
+		break;
+
+	case BT_HAS_OP_SET_PREV_PRESET_SYNC:
+		ret = handle_set_prev_preset(client, &buf, true);
+		break;
+#else
+	case BT_HAS_OP_SET_ACTIVE_PRESET_SYNC:
+	case BT_HAS_OP_SET_NEXT_PRESET_SYNC:
+	case BT_HAS_OP_SET_PREV_PRESET_SYNC:
+		ret = BT_GATT_ERR(BT_HAS_ERR_PRESET_SYNC_NOT_SUPP);
+		break;
+#endif /* CONFIG_BT_HAS_PRESET_SYNC_SUPPORT */
+
+	default:
+		ret = BT_GATT_ERR(BT_HAS_ERR_INVALID_OP);
+	}
+
+	return (ret < 0) ? ret : len;
+}
+
+static void connected(struct bt_conn *conn, uint8_t err)
+{
+	struct has_client *client;
+
+	BT_DBG("conn %p err %d", conn, err);
+
+	if (err != 0 || !bt_addr_le_is_bonded(conn->id, &conn->le.dst)) {
+		return;
+	}
+
+	client = client_get(conn);
+	if (unlikely(!client)) {
+		BT_ERR("Failed to allocate client");
+		return;
+	}
+
+	/* Mark all non-hidden presets to be sent via Preset Changed for bonded device.
+	 *
+	 * XXX: At this point stored GATT CCC configurations are not loaded yet,
+	 *	thus we postpone bt_gatt_is_subscribed check on Control Point CCC to be
+	 *	done in security_changed().
+	 */
+	for (int index = 0; index < ARRAY_SIZE(has_preset_list); index++) {
+		if (has_preset_list[index].hidden) {
+			continue;
+		}
+
+		preset_changed_set(client, index, BT_HAS_CHANGE_ID_GENERIC_UPDATE);
+	}
+}
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	struct has_client *client;
+
+	BT_DBG("conn %p reason %d", conn, reason);
+
+	client = client_find(conn);
+	if (client) {
+		client_free(client);
+	}
+}
+
+static void security_changed(struct bt_conn *conn, bt_security_t level, enum bt_security_err err)
+{
+	struct has_client *client;
+
+	BT_DBG("conn %p level %d err %d", conn, level, err);
+
+	if (err != BT_SECURITY_ERR_SUCCESS) {
+		return;
+	}
+
+	client = client_get(conn);
+	if (unlikely(!client)) {
+		BT_ERR("Failed to allocate client");
+		return;
+	}
+
+	/* Check encryption is enabled */
+	if (level < BT_SECURITY_L2) {
+		return;
+	}
```

Comment:
So these checks if the security level goes from e.g. > L2 to < L2? Can that ever happen? Or is `security_changed` called with L0 or L1 in non-error cases?

Reply:
As far as I remember the security level cannot decrease